docs/proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,189 @@
+    MetricValues []MetricValue `json:"metricValues"`
+}
+
+// a metric value for some object
+type GenericMetrics struct {
+    // the type and object metadata describes the object in question
+    unversioned.TypeMeta `json:",inline"`
+    v1.ObjectMeta        `json:"metadata,omitempty"`
+
+    // indicates the end of the time window containing these metrics (i.e.
+    // these metrics come from some time in [Timestamp-Window, Timestamp])
+    Timestamp unversioned.Time `json:"timestamp"`
+
+    // indicates the duration of the time window containing these metrics
+    Window    unversioned.Duration `json:"window"`
+
+    // the value of the metric for this
+    Value float64
+}
+```
+
+For instance, the example request above would yield the following object:
+
+```json
+{
+    "kind": "Metrics",
+    "apiVersion": "general-metrics/v1alpha1",
+    "metricName": "hits-per-second",
+    "metricValues": [
+        {
+            "kind": "Pod",
+            "metadata": { "name": "server1", "namespace": "webapp" },
+            "timestamp": SOME_TIMESTAMP_HERE,
+            "window": "10s",
+            "value": 10.0
+        },
+        {
+            "kind": "Pod",
+            "metadata": { "name": "server2", "namespace": "webapp" },
+            "timestamp": SOME_TIMESTAMP_HERE,
+            "window": "10s",
+            "value": 15.0
+        }
+    ]
+}
+```
+
+Mechanical Concerns
+-------------------
+
+This API is intended to be implemented by monitoring pipelines (e.g.
+inside Heapster, or as an adapter on top of a solution like Prometheus).
+It shares many mechanical requirements with normal Kubernetes APIs.
+
+The API should support encoding the result in both JSON and Protobuf,
+similarly to how core Kubernetes APIs work.  Most likely, the Protobuf
+encoding will be used by most consumers for efficiency's sake.
+
+Additionally, the API should support encoding the output result in
+different versions, if needed, as is currently the case for normal
+Kubernetes APIs.
+
+### Discoverability ###
+
+While the initial plan for the API is for consumption by the HPA
+controller, the API could be more broadly useful -- custom autoscaler or
+scheduler implementations could make use of it's structured presentation
+of metrics.  Broadly speaking, any consumer that might wish to
+consume "core" metrics could potentially benefit from also being able to
+consume general metrics.
+
+While initially the API seems like a good fit for webhook-style
+consumption (similarly to authentication today, for example), the lack of
+discoverability makes it more difficult for multiple consumers to make use
+of it (each has to be specifically configured to find it at a particular
+location, instead of being able to make use of the normal API discovery
+and summarization plans, as with the master metrics API).
+
+The API should thus be discoverable and accessible as if it were just
+another API server, similarly to the master metrics API.
```

Comment:
I'm fine with having the initial implementation work like this, but I feel that we should eventually move to using discovery, so I think it should be in proposal at some point.